### ISSUE TITLE
Calculation of Green's function in a periodic system

### ISSUE THREAD

**Issue by mahar** [HomerReid/scuff-em#27]
Hello Homer and scuff-em users,

I would like to ask you if it is possible to calculate the Green's function at a point in a periodic medium through the C++ interface in SCUFF-EM. 

Thank you very much!

Charalampos

**Comment by HomerReid** [comment]
Yes, this is possible. The C++ code for this calculation will involve the following steps.

(a) You will instantiate an `RWGGeometry` for a `.scuffgeo` file describing your periodic geometry.

(b) You will create a point dipole source (an instance of `PointSource`) located at the point in question with unit-strength dipole moment in (say) the *x* direction.

(c) For a given frequency and Bloch wavevector, you will solve a scattering problem, in which the incident field is furnished by your `PointSource,` and you will compute the scattered fields (not total fields) at the location of the point source. The *x, y, z* components of the scattered **E** fields will be the elements *G_xx, G_yx, and G_zx* of the electric-electric dyadic Green's functions. (The scattered **H** fields will similar give elements of the electric-magnetic DGFs.)  

(d) However, what you will have thus computed is only  contribution of a single Bloch wavevector to the DGF. (You haven't computed the response of the system to a point source; you have computed the response of the system to a phased array of point sources.) To get the response to an actual point source, you will need to do a numerical integration over the Brillouin zone to sum the contributions of all Bloch wavevectors. 

 I will try to post sample code demonstrating how to do it over the next couple of days.

**Comment by gevero** [comment]
Hi Homer

I am really interested in this topic. Could you point me to some paper or book dealing with the theory described in point (d) of your answer? I struggled a lot with this problem, asking people left and right, but it looks like you know very well this stuff. Thanks a lot.

Best

Giovanni

**Comment by mahar** [comment]
Thank you Homer!

It would be nice to have a sample code.

Charalampos

**Comment by gevero** [comment]
Hi Homer

Any update on this particular problem? I am also extremely interested in this problem, since it is of extreme relevance in plasmonics and nanophotonics

**Comment by HomerReid** [comment]
There has been progress on this, but I haven't had a chance to codify it into a digestible update. I will work on it this weekend and post an update by Tuesday 7/14 at the latest.

**Comment by HomerReid** [comment]
Here are some notes:

http://homerreid.com/scuff-LDOS.pdf

I will post codes that implement the functionality described here by Friday 7/17 at the latest.

**Comment by HomerReid** [comment]
The latest repository commit (231a0d2caec6ee0429554411849bb1806a4da36e) includes a new application code called `scuff-ldos.` This code computes LDOS and dyadic Green's functions for periodic and non-periodic geometries. The documentation for the code is here: 

http://homerreid.dyndns.org/scuff-em/doc/applications/scuff-ldos/scuff-ldos/

and an initial example, somewhat hastily documented, is here:

http://homerreid.dyndns.org/scuff-em/doc/examples/HalfSpaceLDOS/HalfSpaceLDOS/

If this addresses the issue, please close it. (If there are other specific issues related to bugs in this code or other particular considerations regarding it, please create separate issues for those).

**Comment by amrit-poudel** [comment]
Thank you for this update. Sorry for highjacking this thread, but could you please provide the exact analytical expression of the Greens function that you are comparing your numerical solution?

From what I can gather, you are a solving a 3D half space problem. Here the geometry is periodic in x and y directions, but the source is still a point dipole source, not a line source.

**Comment by HomerReid** [comment]
Please see appendix B of the memo (scuff-LDOS.pdf) referenced above.

**Comment by amrit-poudel** [comment]
Thanks for this note. I tried comparing your analytical expression for G_zz (Eq. 11(c) in B.2 section of your note) with previously derived result, http://arxiv.org/abs/quant-ph/9906128, eq. 14 in that paper which is evaluated at source point located z above the half-space.

I could be wrong but I do not see a factor of 2 in the exponent of $E^{+}_TM$ in your expression of scattering field in 11(c). The factor of 2 comes from the fact that the source point and the evaluation point are the same.

**Comment by HomerReid** [comment]
Note that **E**`_{TM}` in equation (11c) is evaluated at the argument **x**=`2 z_0 \hat`{**z**}. Looking at the definition of **E**_{TM} at the beginning of section B.1, you will see that this yields a factor of `\exp(2ikz_0),` corresponding to waves emitted from *-z_0* and propagating to *+z_0*.

**Comment by amrit-poudel** [comment]
Thanks for your reply. In your note, however, the position argument of E^{+}_{TM} in Eq. 11c is 0. I think that is a typo. I think that the 0 in the position argument should be replaced by 2z_0, like you just mentioned. The same is true for 11a and 11b.

**Comment by HomerReid** [comment]
This was an earlier typo that I have since fixed. Is it possible that you are looking at an older version of the memo? I just checked that the online version is the corrected one.

**Comment by amrit-poudel** [comment]
I clicked on the link (http://homerreid.com/scuff-LDOS.pdf) you suggested in one of you posts (above) on this thread. I guess this is linked with an older version. 

Thanks again for clarifying. I appreciate it.

**Comment by HomerReid** [comment]
The link in the **SCUFF-LDOS** documentation page pointed to the correct version of the memo. The older link pointed to an older version. I have now fixed this.
